3D Chess/Assets/Scripts/GameState.cs

```diff
@@ -27,18 +44,49 @@ public void ResetGame()
             Destroy(piece.gameObject);
         }
 
+        // Re-create all RealmObjects with their initial position.
+        pieceEntities = Persistence.ResetDatabase(realm);
+
         // Recreate the GameObjects.
-        pieceSpawner.CreateGameObjects(pieces);
+        CreateGameObjects();
     }
 
     private void Awake()
     {
-        pieceSpawner.CreateGameObjects(pieces);
+        realm = Realm.GetInstance();
+        pieceEntities = realm.All<PieceEntity>();
+
+        // Check if we already have PieceEntity's (which means we resume a game).
+        if (pieceEntities.Count() == 0)
+        {
+            // No game was saved, create the necessary RealmObjects.
+            pieceEntities = Persistence.ResetDatabase(realm);
+        }
+        CreateGameObjects();
+    }
+
+    private void CreateGameObjects()
+    {
+        // Each RealmObject needs a corresponding GameObject to represent it.
+        foreach (PieceEntity pieceEntity in pieceEntities)
+        {
+            PieceType type = (PieceType)pieceEntity.Type;
+            Vector3 position = pieceEntity.GetPosition();
+            pieceSpawner.SpawnPiece(type, position, pieces);
```

Comment:
Should we add an `Entity` field to `Piece` and have the `Piece` object contain its database object? We should probably brainstorm the relationship a little bit, but I'm not a fan of the approach where `GameState` updates both game objects and database objects.

What we could do is have the game object proxy its updates to Realm - e.g. when we update its position, it'll also write the update to its entity, when we destroy it, it will delete the entity and so on.

Reply:
Well, it's basically an interactor between a view and the data. Not saying MVC or MVVM are great but when I tried what you mentioned (which would also be my favorite solution I ran into some deadlocks).

The problem is, if you want to do it that way you would want the `Entity` being created when the `GameObject` is created and the same with deletion.

When setting up the board (or resetting it) that's fine since we start by setting up the `Piece` objects which then handles the rest automatically.

When resuming we don't start with `Piece` but with `PieceEntity`. If I were to create a `Piece` for every `PieceEntity` in the DB that would create another `PieceEntity`. I could actually do a lookup here.

Let me see if the current project structure allows for that. I got a feeling that the changes made during other PRs might have changed something that I didn't see before. 👍

Reply:
After being held up by other things I have now finished that rework. With the other changes we did I found a way to do it as requested and move the `PieceEntity` into `Piece`.

I've also simplified the `GameState` further by extracting code to the `PieceSpawner`. Also, the `Persistence` class is removed and the `Piece` positions are no longer duplicated.

See 89e4224.